### PR TITLE
Support yield in async def callbacks.

### PR DESCRIPTION
**Pull request by wRAR**
Based on #3485. Requires Python 3.6.

I think all results of a callback are only processed when it finishes, but this seems to be the case with non-async callbacks too.

### CODE REVIEW

File: scrapy/utils/spider.py [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
+    from scrapy.utils.py36 import collect_asyncgen
+except SyntaxError:
+    collect_asyncgen = None
 
 
 logger = logging.getLogger(__name__)
 
 
 def iterate_spider_output(result):
+    if collect_asyncgen and hasattr(inspect, 'isasyncgen') and inspect.isasyncgen(result):
```
Comment (kmike) [inline scrapy/utils/spider.py]: Could you please add a comment here, something like "workaround to support async generators in Python 3.6"?

**Comment by kmike** [comment]
Interesting - does it mean the whole thing worked with async generators already in Python 3.7+?

File: tests/py36/_test_crawl.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    def start_requests(self):
+        for i in range(self.initial_reqs):
+            yield self._get_req(i)
+
+    async def parse(self, response):
+        index = response.meta['index']
+        yield {'index': index}
+        if index < 10 ** self.depth:
+            for new_index in range(10 * index, 10 * index + self.following_reqs):
+                yield self._get_req(new_index)
```
Comment (kmike) [inline tests/py36/_test_crawl.py]: For paranoid me: could you please check that non-default callback also works, and that async def works for methods other than parse?

**Comment by kmike** [comment]
@Gallaecio I think Python 3.6 milestone can be removed, because this PR doesn't break compatibility with Python < 3.6.

**Comment by kmike** [comment]
> I think all results of a callback are only processed when it finishes

This is an important limitation, which we should be very upfront about. I'm fine with merging it as-is, as soon as it doesn't make things worse for non-async def callbacks.

> but this seems to be the case with non-async callbacks too.

@wRAR could you please double-check how regular callbacks behave (for requests and for items)?

**Comment by wRAR** [comment]
@kmike for generator callbacks the generator itself is passed to [Scraper.handle_spider_output](https://github.com/scrapy/scrapy/blob/1.8/scrapy/core/scraper.py#L171) where it's iterated and items are passed to `Scraper._process_spidermw_output`. On the way there it's wrapped into a `MutableChain` and into several functions that catch exceptions while iterating it. I'm currently unable to say which parts of this should be changed in order to support passing async generators without turning them to lists first. On the other hand, looks like this PR doesn't support graceful handling of exceptions in async generator callbacks, I'll need to write some tests for that and see what happens.

**Comment by kmike** [comment]
Thanks for the investigation @wRAR. Let's have it in, to make `async def` callbacks slightly easier to use in the meantime.

**Comment by Gallaecio** [comment]
> On the other hand, looks like this PR doesn't support graceful handling of exceptions in async generator callbacks, I'll need to write some tests for that and see what happens.

@wRAR If so, I guess we should mention that in the documentation about asyncio support introduced in https://github.com/scrapy/scrapy/pull/4329

**Comment by wRAR** [comment]
@Gallaecio @kmike I've just checked and if a callback yields an item and then raises an Exception, the item will be produced from the spider if the callback is normal and it won't be produced if it's an async generator.

**Pull request merged**